### PR TITLE
Sk mprk43

### PR DESCRIPTION
**Pull request by SKopecz**
Implementation of MPRK43I and MPRK43II for the solution of a ConservativePDSProblem (currently only out-of-place).

### CODE REVIEW

**Comment by ranocha** [comment]
The spell checker is complaining about the reference "Numer Math" - maybe write out the journal in full?

File: src/mprk.jl [hunk 1]
```diff
@@ -703,40 +703,404 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+
+    # Attention: Implementation assumes that the pds is conservative,
+    # i.e. , P[i, i] == 0 for all i
+
+    # evaluate production matrix
+    P = f.p(uprev, p, t)
+    Ptmp = a21 * P
+    integrator.stats.nf += 1
+
+    # avoid division by zero due to zero patankar weights
```
Comment (ranocha) [inline src/mprk.jl:922]:
```suggestion
    # avoid division by zero due to zero Patankar weights
```

File: src/mprk.jl [hunk 2]
```diff
@@ -703,40 +703,404 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+                alias_A = false, alias_b = false,
+                assumptions = LinearSolve.OperatorAssumptions(true))
+    u2 = sol.u
+    u = u2
+    integrator.stats.nsolve += 1
+
+    # compute Patankar weight denominator
+    σ = σ .^ (1 - q1) .* u .^ q1
+
+    # avoid division by zero due to zero patankar weights
```
Comment (ranocha) [inline src/mprk.jl:941]:
```suggestion
    # avoid division by zero due to zero Patankar weights
```

**Review by ranocha — COMMENTED** [review]
Nice 👍 I just had a quick look (since this is work in progress and the in-place version as well as tests are missing)

File: src/mprk.jl [hunk 3]
```diff
@@ -703,40 +703,404 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+end
+
+function MPRK43I(alpha, beta; thread = False(), linsolve = LUFactorization())
+    MPRK43I{typeof(alpha), typeof(thread), typeof(linsolve)}(alpha, beta,
+                                                             thread, linsolve)
+end
+
+function MPRK43II(gamma; thread = False(), linsolve = LUFactorization())
+    MPRK43II{typeof(gamma), typeof(thread), typeof(linsolve)}(gamma, thread, linsolve)
+end
```
[the remark this answers is not in the document]
Comment (SKopecz) [reply src/mprk.jl:750]: fixed.

File: src/mprk.jl (hunk 1 above)
Comment (SKopecz) [reply src/mprk.jl:922]: fixed.

File: src/mprk.jl (hunk 2 above)
Comment (SKopecz) [reply src/mprk.jl:941]: fixed.

File: src/mprk.jl [hunk 4]
```diff
@@ -703,40 +703,404 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+                alias_A = false, alias_b = false,
+                assumptions = LinearSolve.OperatorAssumptions(true))
+    u = sol.u
+    integrator.stats.nsolve += 1
+
+    # compute Patankar weight denominator
+    if !(q1 ≈ q2)
+        σ = σ0 .^ (1 - q2) .* u2 .^ q2
+
+        # avoid division by zero due to zero patankar weights
```
[the remark this answers is not in the document]
Comment (SKopecz) [reply src/mprk.jl:963]: fixed.

File: src/mprk.jl [hunk 5]
```diff
@@ -703,40 +703,404 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+
+    # solve linear system
+    linprob = LinearProblem(M, uprev)
+    sol = solve(linprob, alg.linsolve,
+                alias_A = false, alias_b = false,
+                assumptions = LinearSolve.OperatorAssumptions(true))
+    σ = sol.u
+    integrator.stats.nsolve += 1
+
+    # avoid division by zero due to zero patankar weights
```
[the remark this answers is not in the document]
Comment (SKopecz) [reply src/mprk.jl:981]: fixed.

[the remark this answers is not in the document]
Comment (ranocha) [reply src/mprk.jl]: Just for your information: You can also click on "Commit suggestion" above to apply the suggested changes here on GitHub. (You need to `git pull` afterwards to get them locally, too)

File: src/mprk.jl [hunk 6]
```diff
@@ -703,40 +703,341 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+    b2 = (3.0 * alg.beta - 2.0) / (6.0 * alg.alpha * (alg.beta - alg.alpha))
+    b3 = (2.0 - 3.0 * alg.alpha) / (6.0 * alg.beta * (alg.beta - alg.alpha))
+    c2 = alg.alpha
+    c3 = alg.beta
+
+    beta2 = 1.0 / (2.0 * a21)
+    beta1 = 1.0 - beta2
+
+    q1 = 1.0 / (3.0 * a21 * (a31 + a32) * b3)
+    q2 = 1.0 / a21
```
Comment (ranocha) [inline src/mprk.jl:829]:
```suggestion
    a21 = alg.alpha
    a31 = (3 * alg.alpha * alg.beta * (1 - alg.alpha) - alg.beta^2) /
          (alg.alpha * (2 - 3 * alg.alpha))
    a32 = (alg.beta * (alg.beta - alg.alpha)) / (alg.alpha * (2 - 3 * alg.alpha))
    b1 = 1 + (2 - 3 * (alg.alpha + alg.beta)) / (6 * alg.alpha * alg.beta)
    b2 = (3 * alg.beta - 2) / (6 * alg.alpha * (alg.beta - alg.alpha))
    b3 = (2 - 3 * alg.alpha) / (6 * alg.beta * (alg.beta - alg.alpha))
    c2 = alg.alpha
    c3 = alg.beta

    beta2 = 1 / (2 * a21)
    beta1 = 1 - beta2

    q1 = 1 / (3 * a21 * (a31 + a32) * b3)
    q2 = 1 / a21
```

This makes it easier to apply the methods with other floating point types such as `Float32`, see https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-using-floats-for-numeric-literals-in-generic-code-when-possible

File: src/mprk.jl [hunk 7]
```diff
@@ -703,40 +703,341 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+    b2 = 3.0 / 4.0 - alg.gamma
+    b3 = alg.gamma
+    c2 = a21
+    c3 = a21
+
+    beta2 = 1.0 / (2.0 * a21)
+    beta1 = 1.0 - beta2
+
+    q1 = 1.0 / (3.0 * a21 * (a31 + a32) * b3)
+    q2 = 1.0 / a21
```
Comment (ranocha) [inline src/mprk.jl:848]:
```suggestion
    a21 = 2 * one(alg.gamma) / 3
    a31 = a21 - 1 / (4 * alg.gamma)
    a32 = 1 / (4 * alg.gamma)
    b1 = one(alg.gamma) / 4
    b2 = 3 * one(alg.gamma) / 4 - alg.gamma
    b3 = alg.gamma
    c2 = a21
    c3 = a21

    beta2 = 1 / (2 * a21)
    beta1 = 1 - beta2

    q1 = 1 / (3 * a21 * (a31 + a32) * b3)
    q2 = 1 / a21
```

File: src/mprk.jl (hunk 6 above)
Comment (SKopecz) [reply src/mprk.jl:829]: Good to know

File: examples/04_example_problemlibrary.jl [hunk 8]
```diff
@@ -1,15 +1,18 @@
 # Install packages
 import Pkg
 Pkg.activate(@__DIR__)
-Pkg.develop(path = dirname(@__DIR__))
-Pkg.instantiate()
+#Pkg.develop(path = dirname(@__DIR__))
+#Pkg.instantiate()
```
[the remark this answers is not in the document]
Comment (SKopecz) [reply examples/04_example_problemlibrary.jl:5]: Yes. Forgot to remove the comment.  The debugger in VS Code stalls if these comments aren't present.

File: src/mprk.jl [hunk 9]
```diff
@@ -703,40 +703,341 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+    M::PType
+    σ::uType
+    tab::tabType
+    thread::Thread
+    linsolve_tmp::uType  # stores rhs of linear system
+    linsolve::F
+    weight::uNoUnitsType
+end
+
+function set_constant_parameters(alg::MPRK43I)
```
Comment (JoshuaLampert) [inline src/mprk.jl:814]:
```suggestion
function get_constant_parameters(alg::MPRK43I)
```
Just a minor thing I noticed while looking at the PR: To me a `set_*` function name sounds like a mutating function, but this rather looks like a `get_*` function to me.

File: src/mprk.jl [hunk 10]
```diff
@@ -703,40 +703,341 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+    beta2 = 1 / (2 * a21)
+    beta1 = 1 - beta2
+
+    q1 = 1 / (3 * a21 * (a31 + a32) * b3)
+    q2 = 1 / a21
+
+    return a21, a31, a32, b1, b2, b3, c2, c3, beta1, beta2, q1, q2
+end
+
+function set_constant_parameters(alg::MPRK43II)
```
Comment (JoshuaLampert) [inline src/mprk.jl:834]:
```suggestion
function get_constant_parameters(alg::MPRK43II)
```
See above.

File: src/mprk.jl [hunk 11]
```diff
@@ -703,40 +703,341 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+    small_constant::T
+end
+
+function alg_cache(alg::Union{MPRK43I, MPRK43II}, u, rate_prototype, ::Type{uEltypeNoUnits},
+                   ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits},
+                   uprev, uprev2, f, t, dt, reltol, p, calck,
+                   ::Val{false}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
+
+    #TODO: Should assert alg.alpha ≠ alg.beta, alg.alpha ≠ 0, alg.beta ≠ 0, alg.alpha ≠ 2/3 
+    a21, a31, a32, b1, b2, b3, c2, c3, beta1, beta2, q1, q2 = set_constant_parameters(alg)
```
Comment (JoshuaLampert) [inline src/mprk.jl:875]:
```suggestion
    a21, a31, a32, b1, b2, b3, c2, c3, beta1, beta2, q1, q2 = get_constant_parameters(alg)
```
See above.

File: examples/04_example_problemlibrary.jl [hunk 12]
```diff
@@ -81,38 +111,74 @@ sims = convergence_tab_plot(prob_pds_nonlinmod,
...
 ## robertson problem ######################################################
 sol_robertson = solve(prob_pds_robertson, Rosenbrock23());
 # Cannot use MPE() since adaptive time stepping is not implemented
-sol_robertson_MPRK = solve(prob_pds_robertson, MPRK22(1.0));
+sol_robertson_MPRK22 = solve(prob_pds_robertson, MPRK22(1.0), reltol = 1e-3, abstol = 1e-2);
+sol_robertson_MPRK43I = solve(prob_pds_robertson, MPRK43I(1.0, 0.5), reltol = 1e-3,
+                              abstol = 1e-2);
+# TODO: Check why MPRK43II is not working
+# TODO: Implement "controller" which doubles timestep size for each step
+sol_robertson_MPRK43II = solve(prob_pds_robertson, MPRK43II(2.0 / 3.0), reltol = 1e-3,
```
[the remark this answers is not in the document]
Comment (SKopecz) [reply examples/04_example_problemlibrary.jl:127]: Only the 2nd is open. Created issue #74.

**Review by ranocha — APPROVED** [review]
Thanks a lot! We can merge this PR from my point of view as it is - but we should remmeber to add some tests to the `test` directory as well. We can either do this in this PR or later - up to you, @SKopecz

**Comment by ranocha** [comment]
Thanks for your review, @JoshuaLampert!

File: src/mprk.jl [hunk 13]
```diff
@@ -703,40 +703,341 @@ function perform_step!(integrator, cache::MPRK22Cache, repeat_step = false)
...
+                                                             thread, linsolve)
+end
+
+"""
+    MPRK43II(γ; [linsolve = ...])
+
+A third-order modified Patankar-Runge-Kutta algorithm for (conservative)
+production-destruction systems. This one-step, four-stage method is
+third-order accurate, unconditionally positivity-preserving, and linearly
+implicit. The parameters `γ` is described by Kopecz and Meister (2018).
```
Comment (JoshuaLampert) [inline src/mprk.jl:748]:
```suggestion
implicit. The parameter `γ` is described by Kopecz and Meister (2018).
```

**Comment by JoshuaLampert** [comment]
I think a short comment in the docstrings of `MPRK43I` and `MPRK43II` explaining the connection between these two schemes would help.

**Comment by SKopecz** [comment]
> I think a short comment in the docstrings of `MPRK43I` and `MPRK43II` explaining the connection between these two schemes would help.

**Comment by SKopecz** [comment]
> Thanks a lot! We can merge this PR from my point of view as it is - but we should remmeber to add some tests to the `test` directory as well. We can either do this in this PR or later - up to you, @SKopecz

I will add the tests.

**Review by ranocha — APPROVED** [review]
Thanks a lot!

**Pull request closed without merging**